File: sample-objc/ARGEARSample/UI/UIComponent/Main/Top/MainTopFunctionView.m
//
//  MainTopFunctionView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "MainTopFunctionView.h"
#import "UIImage+Color.h"

@implementation MainTopFunctionView

#define kMainTopFunctionSettingButtonName @"icSetting"
#define kMainTopFunctionRotateButtonName @"icRotate"

#define kMainTopFunctionRatio11ButtonName @"ic11Black"
#define kMainTopFunctionRatio43ButtonName @"ic43"
#define kMainTopFunctionRatioFullButtonName @"icFullWhite"

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (instancetype)init {
    self = [super init];
    if (self) {
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
    }
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    
    [self setButtons:[self tag]];
}

- (IBAction)settingButtonAction:(UIButton *)sender {
    
    if (_delegate && [_delegate respondsToSelector:@selector(settingButtonAction)]) {
        [_delegate settingButtonAction];
    }
}

- (IBAction)ratioButtonAction:(UIButton *)sender {
    
    if (_delegate && [_delegate respondsToSelector:@selector(ratioButtonAction)]) {
        [_delegate ratioButtonAction];
    }
}

- (IBAction)toggleButtonAction:(UIButton *)sender {
    [self setDisableButtons];
    
    if (_delegate && [_delegate respondsToSelector:@selector(toggleButtonAction)]) {
        [_delegate toggleButtonAction];
    } else {
        [self setEnableButtons];
    }
}

- (void)setEnableButtons {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [[self settingButton] setUserInteractionEnabled:YES];
        [[self ratioButton] setUserInteractionEnabled:YES];
        [[self toggleButton] setUserInteractionEnabled:YES];
    });
}

- (void)setDisableButtons {
    [_settingButton setUserInteractionEnabled:NO];
    [_ratioButton setUserInteractionEnabled:NO];
    [_toggleButton setUserInteractionEnabled:NO];
}

- (void)setButtons:(ARGMediaRatio)ratio {
    
    CGFloat screenRatio = [[UIScreen mainScreen] bounds].size.height / [[UIScreen mainScreen] bounds].size.width;

    NSMutableString *settingButtonImageString = [NSMutableString stringWithString:kMainTopFunctionSettingButtonName];
    NSMutableString *rotateButtonImageString = [NSMutableString stringWithString:kMainTopFunctionRotateButtonName];
    NSMutableString *ratioButtonImageString = [NSMutableString stringWithString:@""];
    
    switch (ratio) {
        case ARGMediaRatio_16x9:
            [settingButtonImageString appendString:@"White"];
            [rotateButtonImageString appendString:@"White"];
            [ratioButtonImageString appendString:kMainTopFunctionRatioFullButtonName];
            break;
        case ARGMediaRatio_4x3:
            if (screenRatio > 2.0f) {
                [settingButtonImageString appendString:@"Black"];
                [rotateButtonImageString appendString:@"Black"];
                [ratioButtonImageString appendString:kMainTopFunctionRatio43ButtonName];
                [ratioButtonImageString appendString:@"Black"];
            } else {
                [settingButtonImageString appendString:@"White"];
                [rotateButtonImageString appendString:@"White"];
                [ratioButtonImageString appendString:kMainTopFunctionRatio43ButtonName];
                [ratioButtonImageString appendString:@"White"];
            }
            break;
        case ARGMediaRatio_1x1:
            [settingButtonImageString appendString:@"Black"];
            [rotateButtonImageString appendString:@"Black"];
            [ratioButtonImageString appendString:kMainTopFunctionRatio11ButtonName];
            break;
        default:
            break;
    }
    
    [_settingButton setImage:[UIImage imageNamed:[NSString stringWithString:settingButtonImageString]] forState:UIControlStateNormal];
    [_settingButton setImage:[[UIImage imageNamed:[NSString stringWithString:settingButtonImageString]] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];

    [_ratioButton setImage:[UIImage imageNamed:[NSString stringWithString:ratioButtonImageString]] forState:UIControlStateNormal];
    [_ratioButton setImage:[[UIImage imageNamed:[NSString stringWithString:ratioButtonImageString]] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];

    [_toggleButton setImage:[UIImage imageNamed:[NSString stringWithString:rotateButtonImageString]] forState:UIControlStateNormal];
    [_toggleButton setImage:[[UIImage imageNamed:[NSString stringWithString:rotateButtonImageString]] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];
}

# pragma mark - Observers
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    
    if ([keyPath isEqualToString:@"currentRatio"]) {
        ARGMediaRatio ratio = [[change objectForKey:@"new"] integerValue];
        
        [self setButtons:ratio];
    }
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/Main/Top/MainTopFunctionView.swift
//
//  MainTopFunctionView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/29.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit
import ARGear

protocol MainTopFunctionDelegate: class {
    func settingButtonAction()
    func ratioButtonAction()
    func toggleButtonAction()
}

class MainTopFunctionView: UIView {
    
    let kMainTopFunctionSettingButtonName = "icSetting"
    let kMainTopFunctionToggleButtonName = "icRotate"

    let kMainTopFunctionRatio11ButtonName = "ic11Black"
    let kMainTopFunctionRatio43ButtonName = "ic43"
    let kMainTopFunctionRatioFullButtonName = "icFullWhite"
    
    weak var delegate: MainTopFunctionDelegate?

    @IBOutlet weak var settingButton: UIButton!
    @IBOutlet weak var ratioButton: UIButton!
    @IBOutlet weak var toggleButton: UIButton!
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.setRatio(ARGMediaRatio(rawValue: self.tag) ?? ._4x3)
    }
    
    func setRatio(_ ratio: ARGMediaRatio) {
        
        let screenRatio = UIScreen.main.bounds.height / UIScreen.main.bounds.width
        
        var settingButtonImageName = self.kMainTopFunctionSettingButtonName
        var toggleButtonImageName = self.kMainTopFunctionToggleButtonName
        var ratioButtonImageName = ""

        switch ratio {
        case ._16x9:
            settingButtonImageName.append("White")
            toggleButtonImageName.append("White")
            ratioButtonImageName.append(self.kMainTopFunctionRatioFullButtonName)
        case ._4x3:
            if screenRatio > 2.0 {
                settingButtonImageName.append("Black")
                toggleButtonImageName.append("Black")
                ratioButtonImageName.append(self.kMainTopFunctionRatio43ButtonName)
                ratioButtonImageName.append("Black")
            } else {
                settingButtonImageName.append("White")
                toggleButtonImageName.append("White")
                ratioButtonImageName.append(self.kMainTopFunctionRatio43ButtonName)
                ratioButtonImageName.append("White")
            }
        case ._1x1:
            settingButtonImageName.append("Black")
            toggleButtonImageName.append("Black")
            ratioButtonImageName.append(self.kMainTopFunctionRatio11ButtonName)
        default:
            break
        }
        
        self.settingButton.setImage(UIImage(named: settingButtonImageName), for: .normal)
        self.settingButton.setImage(UIImage(named: settingButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)
        
        self.ratioButton.setImage(UIImage(named: ratioButtonImageName), for: .normal)
        self.ratioButton.setImage(UIImage(named: ratioButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)

        self.toggleButton.setImage(UIImage(named: toggleButtonImageName), for: .normal)
        self.toggleButton.setImage(UIImage(named: toggleButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)
    }
    
    @IBAction func settingButtonAction(_ sender: UIButton) {
        guard let delegate = self.delegate else { return }
        
        delegate.settingButtonAction()
    }
    
    @IBAction func ratioButtonAction(_ sender: UIButton) {
        guard let delegate = self.delegate else { return }

        delegate.ratioButtonAction()
    }
    
    @IBAction func toggleButtonAction(_ sender: UIButton) {
        self.disableButtons()

        guard let delegate = self.delegate
            else {
                self.enableButtons()
                return
        }

        delegate.toggleButtonAction()
    }
    
    func enableButtons() {
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
            self.settingButton.isUserInteractionEnabled = true
            self.ratioButton.isUserInteractionEnabled = true
            self.toggleButton.isUserInteractionEnabled = true
        }
    }
    
    func disableButtons() {
        self.settingButton.isUserInteractionEnabled = false
        self.ratioButton.isUserInteractionEnabled = false
        self.toggleButton.isUserInteractionEnabled = false
    }
}
